daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Error.scala

```diff
@@ -4,73 +4,142 @@
 package com.daml.lf
 package engine
 
-import com.daml.lf.transaction.GlobalKey
-import com.daml.lf.value.Value._
-
-//TODO: Errors
-sealed trait Error {
-  // msg is intended to be a single line message
-  def msg: String
-
-  // details for debugging should be included here
-  def detailMsg: String
-  override def toString: String = "Error(" + msg + ")"
+import com.daml.lf.data.Ref
+import com.daml.lf.transaction.{GlobalKey, GlobalKeyWithMaintainers}
+import com.daml.lf.value.Value
+
+sealed abstract class Error {
+  def subError: Error.SubError
+  final def msg: String = subError.msg
+  final def detailMsg: String = msg
 }
 
 object Error {
 
-  def withError[A](e: Error)(optional: Option[A]): Either[Error, A] = {
-    optional.fold[Either[Error, A]](Left(e))(v => Right(v))
+  sealed abstract class SubError extends RuntimeException with scala.util.control.NoStackTrace {
+    // msg is intended to be a single line message
+    def msg: String
+    // details for debugging should be included here
+    def detailMsg: String
+    def toError: Error
   }
 
-  /** small conversion from option to either error with specific error */
-  implicit class optionError[A](o: Option[A]) {
-    def errorIfEmpty(e: Error) = withError(e)(o)
+  // Error happening during Package loading
+  final case class Package(subError: Package.SubError) extends Error
+
+  object Package {
+
+    sealed abstract class SubError extends Error.SubError {
+      final def toError: Package = Package(this)
+      override def detailMsg: String = msg
+    }
+
+    // TODO: get ride of Generic
```

Comment:
```suggestion
    // TODO: get rid of Generic
```
sounds more french with an `e` but I think the English spelling doesn’t have one :)